@xen-orchestra/xapi/vdi.mjs

```diff
@@ -95,96 +96,114 @@ class Vdi {
+      baseParentUuid // the uuid of the parent used as a base for a delta export
+
+    if (baseRef !== undefined && preferNbd && cbt_enabled) {
+      // use CBT if possible
+      // call to list changed blocks must be done before the vdi is mounted for NBD export
+      try {
+        changedBlocks = await this.VDI_listChangedBlock(ref, baseRef)
+        baseParentUuid = await this.getField('VDI', baseRef, 'sm_config').then(sm_config => sm_config?.['vhd-parent'])
+
+        info('found changed blocks', changedBlocks)
+      } catch (error) {
+        // do not fail if CBT is not enabled/working
+        info('no changed block', error)
       }
-      // the raw nbd export does not need to peek ath the vhd source
-      if (nbdClient !== undefined && format === VDI_FORMAT_RAW) {
-        stream = createNbdRawStream(nbdClient)
-      } else {
-        if (changedBlocks === undefined) {
-          // raw export without nbd or vhd exports needs a resource stream
-          // also metadata vdis (vdi that have a data_destroy) can't export a stream
-          stream = exportStream = (
-            await this.getResource(cancelToken, '/export_raw_vdi/', {
-              query,
-              task: await this.task_create(`Exporting content of VDI ${vdiName}`),
-            })
-          ).body
-        }
+    }
 
-        if (nbdClient !== undefined && format === VDI_FORMAT_VHD) {
-          const taskRef = await this.task_create(
-            `Exporting content of VDI ${vdiName} using NBD ${changedBlocks !== undefined ? ' and CBT' : ''}`
-          )
-          exportStream = stream
-          stream = await createNbdVhdStream(nbdClient, exportStream, {
-            changedBlocks,
-            vdiInfos: { size, uuid, parentUuid: baseParentUuid },
-          })
-          stream.on('progress', progress => this.call('task.set_progress', taskRef, progress))
-          finished(stream, () => {
-            nbdClient.disconnect()
-            exportStream?.destroy() // ensure the source stream is really closed
-
-            this.task_destroy(taskRef).catch(warn)
-          })
+    // really connect to NBD server
+    if (preferNbd) {
+      nbdClient = await this._getNbdClient(ref, { nbdConcurrency })
+      // disconnect on failure or when transfer is finished
+      $defer.onFailure(() => nbdClient?.disconnect())
+    }
+
+    // create a xapi export stream only if we won't make a delta from CBT
+    // a CBT export can only work if we have a NBD client and changed blocks
+    if (changedBlocks === undefined || nbdClient === undefined) {
+      stream = exportStream = (
+        await this.getResource(cancelToken, '/export_raw_vdi/', {
+          query,
+          task: await this.task_create(`Exporting content of VDI ${vdiName}`),
+        })
+      ).body
+      exportStream.on('error', err => {
+        // this error code is directly caused by exportStream.destroy
+        if (err.code !== 'UND_ERR_ABORTED') {
+          throw err
```

Comment:
Throwing an error in an even listener is generally forbidden (will break other event listeners and nobody can catch it.

Either ignore it or log it.